### PR TITLE
enable modifying exercises

### PR DESCRIPTION
**Pull request by jelmnainen**

### CODE REVIEW

File: src/components/ExercisesEditPage.jsx [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      exercise.get('id'),
+      title.value,
+      text.value,
+      category.value,
+      batch.value,
+      fileUpload.checked
+    );
+  }
+
+  renderOptionItems(set, chosenMember, refName) {
```
Comment (nfrolov) [inline src/components/ExercisesEditPage.jsx]: 1. rename `renderOptionItems` to `renderSelect`
2. print field's label
3. `chosenMember` could be simple id as a `defaultValue`

File: src/components/ExercisesEditPage.jsx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    return (
+      <div className="field">
+        <select className="ui dropdown" ref={refName}>
```
Comment (nfrolov) [inline src/components/ExercisesEditPage.jsx]: Print all options and use `defaultValue` on the select.

File: src/components/ExercisesEditPage.jsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            .valueSeq()
+            .filter(member => member.get('id') !== chosenMember.get('id'))
+            .map(member =>
+              <option key={member.get('id')} value={member.get('id')}>
+                {member.get('title')}
+              </option>
+            )
+          }
+        </select>
+      </div>
```
Comment (nfrolov) [inline src/components/ExercisesEditPage.jsx:55]: Possible error messages are not printed. (batch is required)

File: src/components/ExercisesEditPage.jsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  render() {
+    const { exercise, categories, batches, form, userIsAllowedToEdit } = this.props;
+
+    if (!exercise || !categories || !batches || !form) {
+      return (
+        <h3>Loading...</h3>
+      );
+    }
+
+    const errorMessages = form.get('errorMessages');
```
Comment (nfrolov) [inline src/components/ExercisesEditPage.jsx]: `form.get('errorMessages', Map())`

File: src/components/ExercisesEditPage.jsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    if (!userIsAllowedToEdit) {
+      return (
+        <h3>
+          You don't have the rights to edit this exercise
+        </h3>
+      );
+    }
+
+    if (errorMessages && errorMessages.get('title') && errorMessages.get('title').count() > 0) {
```
Comment (nfrolov) [inline src/components/ExercisesEditPage.jsx]: `if (errorMessages.get('title', List()).count())`

File: src/actions/exercisesActions.js [hunk 6]
```diff
@@ -95,3 +100,27 @@ export const deleteExercise = (id) => ({
     method: 'delete',
   },
 });
+
+export const patchExercise = (id, title, text, categoryID, batchID, fileUpload) => ({
```
Comment (nfrolov) [inline src/actions/exercisesActions.js]: `ID` should be `Id`.
I would use `(id, exercise)` definition where `exercise` is an object.

File: src/containers/ExercisesEditContainer.jsx [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  categories: state.getIn(['categories', 'entries']),
+  batches: state.getIn(['batches', 'entries']),
+  userIsAllowedToEdit: state.getIn(['auth', 'isTeacher']) && state.getIn(['auth', 'isSignedIn']),
+  form: state.getIn(['exercises', 'updateForm']),
+});
+
+
+const mapDispatchToProps = (dispatch) => ({
+  fetchExercise(exerciseID) {
+    dispatch(fetchCategories());
```
Comment (nfrolov) [inline src/containers/ExercisesEditContainer.jsx:21]: `Id`

File: src/components/ExercisesEditPage.jsx (hunk 3 above)
Comment (jelmnainen) [inline src/components/ExercisesEditPage.jsx:55]: It's not possible to select nothing in this dropdown

Comment (nfrolov) [inline src/components/ExercisesEditPage.jsx:55]: That is a problem too. First, category is an optional. Second, we have an implicit default category/batch.
Empty option with empty value should be provided.

File: src/components/ExercisesEditPage.jsx [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+      return (
+        <h3>Loading...</h3>
+      );
+    }
+
+    const errorMessages = form.get('errorMessages', Map());
+    const isUpdated = form.get('isCreated');
+    const currentCategoryId = categories.getIn([exercise.get('category_id')]).get('id');
+    const currentBatchId = batches.getIn([exercise.get('batch_id')]).get('id');
+
```
Comment (nfrolov) [inline src/components/ExercisesEditPage.jsx:70]: Get id by id :-)

File: src/components/ExercisesEditPage.jsx [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+          You don't have the rights to edit this exercise
+        </h3>
+      );
+    }
+
+    let titleErrors;
+    let textErrors;
+    let updated;
+
+    if (errorMessages && errorMessages.get('title', List()).count() > 0) {
```
Comment (nfrolov) [inline src/components/ExercisesEditPage.jsx]: `errorMessages` is now always defined.

File: src/actions/exercisesActions.js [hunk 10]
```diff
@@ -95,3 +100,27 @@ export const deleteExercise = (id) => ({
...
+      EXERCISES_UPDATE_SUCCESS,
+      EXERCISES_UPDATE_FAILURE,
+    ],
+    endpoint: `/exercises/${id}`,
+    method: 'put',
+    body: {
+      title,
+      text,
+      category_id: categoryId,
+      batch_id: batchId,
```
Comment (nfrolov) [inline src/actions/exercisesActions.js:117]: Now these can be shortened. Will be converted to snake_case.

Comment (jelmnainen) [inline src/actions/exercisesActions.js:117]: Not personally a fan of magicks

**Pull request merged**